sycl/include/CL/sycl/detail/pi.h

```diff
@@ -1600,6 +1600,11 @@ __SYCL_EXPORT pi_result piextUSMGetMemAllocInfo(
     pi_context context, const void *ptr, pi_mem_info param_name,
     size_t param_value_size, void *param_value, size_t *param_value_size_ret);
 
+/// API to get Plugin/Device internal data to be directly used by
+/// kernel implementation directly
+__SYCL_EXPORT pi_result piextPluginGetOpaqueData(void *opaque_data_param,
+                                                 void **opaque_data_return);
```

Comment:
```suggestion
/// This structure denotes a plugin-specific data returned via the
/// piextPluginGetOpaqueData PI call. Depending on the \c version field, the second \c data
/// field is can be interpreted differently. Versioning and data interpretation is plugin-specific.
struct PluginOpaqueData {
  uintptr_t version;
  void *data;
};
/// API to get Plugin internal data, opaque to SYCL RT. Some devices whose device code is compiled
/// by the host compiler (e.g. CPU emulators) may use it to access some device code functionality
/// implemented in/behind the plugin.
/// \param opaque_data_param - unspecified argument, interpretation is specific to a plugin
/// \param opaque_data_return - placeholder for the returned opaque data.
__SYCL_EXPORT pi_result piextPluginGetOpaqueData(void *opaque_data_param,
                                                 PluginOpaqueData **opaque_data_return);
```

Reply:
Why would we mandate the versioning at PI level? The data is already "opaque", so plugins are free to add version there if wanted.

Reply:
@smaslov-intel - my recollection is that it was suggested by someone during our internal discussion, and this was adopted. I'm fine to re-consider and represent it as an opaque pointer on PI level.

Reply:
It wasn't me, and I'd prefer not to have any protocol for this opaque data.

Reply:
It seems that Konst refers to what I mentioned in the discussion. What I meant was that we have to separate PI_API version (`PiVersion` from `struct _pi_plugin` defined `pi.h`) and version of opaque data itself (`version` from `struct PluginOpaqueData`).

Reply:
I agree that `PiVersion` should not be connected to the version of data in `piextPluginGetOpaqueData`. But I also suggest that we don't mandate in PI that that opaque data has a version. It would be up to individual plugin to decide what to put in the data.

Reply:
@smaslov-intel , do you mean that the data-versioning should not be exposed at the PI level while being hidden behind the `void*` type pointer? That was what I originally intended with `void **opaque_data_return`.

Reply:
Yes, I like that version better.

Reply:
@smaslov-intel , @kbobrovs 

I would like to finalize this API definition and `void*` vs `PluginOpaqueData` in `pi.h`. Please share your opinion.

Reply:
> I'm fine to re-consider and represent it as an opaque pointer on PI level.

I read this as we are in agreement to proceed with `void**` version.

Reply:
Okay.